### ISSUE TITLE
Panning is too slow

### ISSUE THREAD

**Issue by NarasimhaShenoi** [cerner/carbon-graphs#311]
**What part of Carbon does this bug correlate to?**
General graph logic:

* [X]  Graph attributes (axes, legends, regions...etc.)
* [ ]  Other (please specify): 

Specific graph logic:

* [ ]  Bar graph
* [ ]  Bubble graph
* [ ]  Gantt chart
* [ ]  Line graph
* [ ]  Paired result
* [ ]  Pie chart
* [ ]  Scatter plot
* [ ]  Timeline

**Describe the bug**
When we try to load a graph during reflow it takes a considerable amount of time than the time taken to create the graph. 

**Expected behavior**
the time to load the graph for that particular timeframe during reflow should be less than the time taken to create the graph since we are not destroying or the existing graph object and we are just updating the graph.
It took for hardly 4 seconds to load the graph
<img width="1475" alt="Screenshot 2020-10-19 at 2 40 09 PM" src="https://user-images.githubusercontent.com/53594676/96424903-00183180-1219-11eb-9f8d-db63bd9a7184.png">

Whereas it took around 7 seconds to update graph during panning
<img width="1506" alt="Screenshot 2020-10-19 at 2 44 49 PM" src="https://user-images.githubusercontent.com/53594676/96425437-a7956400-1219-11eb-837c-d3403eea0b31.png">

**Comment by sdadn** [comment]
Can you include a code sandbox that demonstrates this behavior?

**Comment by NarasimhaShenoi** [comment]
@sdadn The reflow legend takes more time when there are multiple legends in the graph canvas. I am have made a sample [code sandbox](https://codesandbox.io/s/long-snowflake-vjf2f?file=/src/graph.js) that shows the same behaviour.

**Comment by sdadn** [comment]
@NarasimhaShenoi in the sandbox, why are you calling the reflow method separately for each dataset?
```js
const createGraph = () => {
  graph.config.axis.x.lowerLimit = new Date(2016, 0, 1, 23, 59).toISOString();
  graph.config.axis.x.upperLimit = new Date(2016, 0, 0, 11, 59).toISOString();
  graph.reflow({ ...CONTENT_DATA });
  graph.reflow({ ...CONTENT_DATA2 });
  graph.reflow({ ...CONTENT_DATA3 });
  graph.reflow({ ...CONTENT_DATA4 });
  graph.reflow({ ...CONTENT_DATA4 });
  graph.reflow({ ...CONTENT_DATA6 });
  graph.reflow({ ...graphData1 });
  graph.reflow({ ...graphData2 });
  graph.reflow({ ...graphData3 });
  graph.reflow({ ...graphData4 });
  graph.reflow({ ...graphData5 });
  graph.reflow({ ...graphData6 });
};
```
According to the documentation, the reflow method should be called once and additional datasets should be loaded usin the `loadContent` method:

https://github.com/cerner/carbon-graphs/blob/main/docs/controls/Panning.md

```js
export const renderLineWithPanning = (id) => {
    const axisData = graphConfiguration;
    axisData.pan = {
        enabled: true
    };
    const graphData = dataSet;
    graphData.regions = [regions[0]];

    const createGraph = () => {
        graph.reflow();
    };

    const graph = Carbon.api.graph(axisData);
    graph.loadContent(Carbon.api.line(graphData));
    // Additional data-sets to be loaded here only, like:
    graph.loadContent(Carbon.api.line(/* Data array */));
    axisData.axis = graph.config.axis;

    createPanningControls(id, {
        axisData,
        creationHandler: createGraph
    });
    return graph;
};
```

**Comment by NarasimhaShenoi** [comment]
@sdadn loadContent() is used to load the data series(or legend) into the graph object. For example, in the dev site, we will load the content for each line graph [here](https://github.com/cerner/carbon-graphs/blob/main/dev/examples/controls/line.js#L561). We need to call the reflow method separately for each dataset.

**Comment by sdadn** [comment]
@NarasimhaShenoi I tried a few examples from the dev site and your sandbox and I'm not seeing any difference between calling the reflow method once or separately per dataset, besides performance. Unless if you have an example that specifically shows why calling the reflow method separately is required, this does not seem to be a bug for general use cases. The examples and documentation both indicate to call it once.

**Comment by pranav300** [comment]
@sdadn
We need to call the reflow function for each dataset that needs to be updated (in the sense that some new data is being added to the dataset with the key already present in the graph) for that particular timeframe.
For eg., if there are 2 datasets for blood pressure in a graph, and the user requires to dynamically update both the datasets once the panning button is clicked. Then they must call reflow twice once for each dataset.

I have created a [sandbox](https://codesandbox.io/s/prod-currying-g17zz) to view what happens when you use loadContent instead of reflow for multiple datasets. View `line 16` in graph.js.

And regarding the performance issue, I think we can improve it by changing the graphData sent by the user into an object where users can send in multiple key-value pairs, and later we can filter out which graph type to call over [here](https://github.com/cerner/carbon-graphs/blob/main/src/main/js/controls/Graph/Graph.js#L437). This will save us the time required to check whether the graph axes has been updated or not. And reduce the reflow calls from users' end.

**Comment by NarasimhaShenoi** [comment]
@sdadn / @mraplumb / @AshishMotanamGurunadham  If this can be picked up, it will very helpful for us. it will increase the performance to a considerable amount.

**Comment by sdadn** [comment]
Hi @NarasimhaShenoi , sorry we've had our hands full with some recent issues that came in. But this is still on the table and we are exploring potential solutions.

**Comment by sdadn** [comment]
Hello @NarasimhaShenoi , we are looking at a solution to improve performance and consumer experience but it is a long term solution that won't be implemented immediately. In the meantime, here is one potential short-term solution:

There was a recent update that improved the performance of loading and unloading datasets by passing them as an array:

```js
graph.loadContent([dataset1, dataset2]);
```
Similarly, you can unload content this way:

```js
graph.unloadContent([dataset1, dataset2]);
```

My suggestion for now would be to try to unload and load datasets within the panning button method. You can also add logic to make sure only appropriate datasets are unloaded and loaded with new data. 


```js
const createGraph = () => {
  graph.config.axis.x.lowerLimit = new Date(2016, 0, 1, 23, 59).toISOString();
  graph.config.axis.x.upperLimit = new Date(2016, 0, 0, 11, 59).toISOString();
  graph.unloadContent(old_dataset_array);
  graph.loadContent(new_dataset_array);
  graph.reflow();
};
```
Let us know if you have any questions or if this solution doesn't work, then we can look at other short-term options.

**Comment by NarasimhaShenoi** [comment]
Hey @sdadn, as you suggested I tried to unload and load the datasets within the panning button, but I was getting this error: Error: Invalid input format, unique key must be provided. I think it might be result of same input key we are trying to pass via panning. Not sure, I am doing it wrongly. Can you please provide me a sandbox example. Also, I am assuming, if we are passing the dataset as a batch, then we will not be using the changes we did in the reflow.

**Comment by sdadn** [comment]
@NarasimhaShenoi  Yes, the error, `Error: Invalid input format, unique key must be provided` occurs when you’re loading two datasets with the same key.  So you’ll need additional logic on your end to make sure that you are not loading a dataset with an existing key. Two possible solutions are:
* Make sure the keys are unique, such as `DatasetA_1`, `DatasetA_2` etc.
* Or maintain an array/object that keeps tracks of what datasets are loaded/unloaded.

Hope this helps. Let me know if you're still having trouble.

**Comment by NarasimhaShenoi** [comment]
@sdadn  Can you please provide a sandbox example for the same? I am not sure how we can ensure the reflow concept will be working if we load and unload content with uniques keys as you suggested. During reflow, the legend state will be retained as it is. But with the approach of loading and unloading the dataset, I don't think it can be retained. 

I believe the [approach](https://github.com/cerner/carbon-graphs/issues/311#issuecomment-722207539) mentioned by Pranav above is a good one where we pass multiple key-value pairs as an array to reflow and do the logic in carbon reflow.  Let me know your thoughts

**Comment by pranav300** [comment]
I have created a [branch](https://github.com/cerner/terra-graphs/tree/optimize-panning) based on the solution that I had proposed earlier [over here](https://github.com/cerner/carbon-graphs/issues/311#issuecomment-722207539).

I have tested this on @NarasimhaShenoi repo and the result is as follows:

With the current method: Approximately 8-10s
<img width="1792" alt="Screenshot 2021-01-07 at 1 47 09 PM" src="https://user-images.githubusercontent.com/29273628/103924166-afd7ae00-513b-11eb-9012-e80838ad609f.png">

With the new changes: Approximately 2-2.5s
<img width="1792" alt="Screenshot 2021-01-07 at 1 41 10 PM" src="https://user-images.githubusercontent.com/29273628/103924243-c5e56e80-513b-11eb-8550-86f0309972d7.png">

Currently, I have only made changes for the graph construct as a POC, based on which I can update Timeline and Gantt as well.

cc/ @mjhenkes @mraplumb @sdadn
